src/shared/calendarspec.c

```diff
@@ -1320,6 +1324,14 @@ static int find_next(const CalendarSpec *spec, struct tm *tm, usec_t *usec) {
                 *usec = tm_usec;
                 return 0;
         }
+
+        /* It seems we entered an infinite loop. Let's gracefully return an error instead of hanging or
+         * aborting. This code is also exercised when timers.target is brought up during early boot, so
+         * aborting here is problematic and hard to diagnose for users. */
+        _cleanup_free_ char *s = NULL;
+        (void) calendar_spec_to_string(spec, &s);
+        return log_warning_errno(SYNTHETIC_ERRNO(EDEADLK),
```

Comment:
I think LOG_DEBUG should be enough here. I mean, the error is propagated up here which should cause a logged about failure anyway, and the error is pretty descriptive even. And debug logging should give then enough hints

Reply:
My idea was that this *shouldn't* happen. I.e. if we get here it means that we have an error in our logic. Hence the  high log level.

Reply:
But then an assert() should be good too?

Reply:
Please see the big comment above: `assert` would lead to boot failure and would be very unpleasant to users. An error here *shouldn't* happen, but it has already happened a few times in the past, so I think it's reasonable to look for a solution that makes it easy for users to diagnose the issue and report to us when it happens again. Part of the problem is that we depend on the details of mktime() here, and even if our code works fine, glibc may be updated and something strange might happen in the future. Also, people are creative with timezones, so maybe some future combination of dst and leap second changes will trip us again. A warning instead of a failure is much more user friendly.